### ISSUE TITLE
No connectivity, when only some nodes are running

### ISSUE THREAD

**Issue by ehlers** [GNS3/gns3-gui#2177]
GNS3 version 2.0.4dev1 on Darwin (64-bit) with Python 3.5.2 Qt 5.6.0 and PyQt 5.6, official GNS3 VM with latest 2.0.4dev1.

Topology, all nodes (including switch) are running in the GNS3VM:
![test_project](https://user-images.githubusercontent.com/896406/28456406-7a748ad2-6e02-11e7-80a9-6a8e51b8811c.png)

When I only start R1 and alpine-2, leaving alpine-1 stopped, I'm unable to communicate between R1 and alpine-2. alpine-2 shows only one received packet, R1 no received packets.

As soon as I start alpine-1 the full connectivity between all nodes is available. After stopping alpine-1 still connectivity between R1 and alpine-2. After stopping all nodes and then starting only R1 and alpine-2 again no connectivity.

Here the topology file (with .txt extension added):
[test.gns3.txt](https://github.com/GNS3/gns3-gui/files/1164967/test.gns3.txt)

**Comment by noplay** [comment]
I think the ethernet switch from dynamips is buggy. I have seen similar issue when playing with vpcs. 

Can you try:
* Start alpine2 and R1
* Ping alpine2 from R1 (will faill)
* Ping R1 from alpine2 (will faill)
* Ping alpine2 from R1 : It's work?

**Comment by ehlers** [comment]
This retested. Now I have the issue between R1 and alpine-1, When starting alpine-2 the connectivity resumes.

To your question (now testing with alpine-1):
* Start alpine1 and R1
* Ping alpine1 from R1: fails
* Ping R1 from alpine1: fails
* Ping alpine1 from R1 : fails

**Comment by grossmj** [comment]
I am suspecting the switch too.

What are the MAC addresses for all these nodes?

**Comment by ehlers** [comment]
Now I have the issue again between alpine-2 and R1

alpine-1: 7A:79:E9:CF:B0:0A
alpine-2: AE:7D:41:7B:BE:9C
R1: c201.0a0d.0000

The MAC address of R1 is stable, the docker MAC addresses change every time the project is loaded.

**Comment by noplay** [comment]
Mac address are randomly assigned by the kernel when we create the veth

**Comment by ehlers** [comment]
I had the idea, that dynamips is blocking in a packet send or receive system call, but that's not the case. I've enhanced the dynamips nio debug to show some message before and after a send/receive system call, but those calls are not blocking.

Here the dynamips nio debug log attached:
[dynamips_log.txt](https://github.com/GNS3/gns3-gui/files/1166135/dynamips_log.txt)

As soon the switch gets some packets from the previous inactive link, the switch continues to run normally.

**Comment by ehlers** [comment]
Funny, when I move the switch to the Mac, leaving the other nodes on the GNS3VM, then it's working fine. Seems to be linux specific.

**Comment by ehlers** [comment]
Found it !!!

My idea of blocking was correct, I just got bitten by the buffering of stdout. By using (the unbuffered) stderr the log shows, that the last action is a read request, that blocks, see [dynamips_fail.txt](https://github.com/GNS3/gns3-gui/files/1167208/dynamips_fail.txt)

I assume the following is happening:

When dynamips sends a packet to an UDP socket, where the other end is not established, select() tells dynamips, that there is data on this socket and a read() return the error "Connection refused".

But sometimes select() returns two file descriptor, one with data und the unconnected one with the error state. When the one with data is processed first and it's a broadcast (or unknown unicast), then it sends the packet also to the unconnected socket. As the error state on this socket is set, the write() returns an error and clears the error state. As select() returned also the unconnected socket, dynamips starts a read() on it. As the error state on this socket is already delivered and cleared, this read blocks.

My fix is to enable nonblocking mode on UDP sockets in dynamips:
```
diff --git a/common/net.c b/common/net.c
index f5ef82b..c87f363 100644
--- a/common/net.c
+++ b/common/net.c
@@ -8,6 +8,7 @@
 #include "utils.h"
 #include "net.h"
 #include "crc.h"
+#include <fcntl.h>

 /*
  * IP mask table, which allows to find quickly a network mask
@@ -342,6 +343,14 @@ int udp_connect(int local_port,char *remote_host,int remote
    }

    freeaddrinfo(res0);
+
+   /* set to nonblocking */
+   if (sck >= 0) {
+      int flags = fcntl(sck, F_GETFL, 0);
+      if (flags >= 0)
+         fcntl(sck, F_SETFL, flags | O_NONBLOCK);
+   }
+
    return(sck);
 }
 #else
@@ -387,8 +396,14 @@ int udp_connect(int local_port,char *remote_host,int remote
    if (connect(sck,(struct sockaddr *)&sin,sizeof(sin)) < 0) {
       perror("udp_connect: connect");
       close(sck);
+      return(-1);
    }

+   /* set to nonblocking */
+   int flags = fcntl(sck, F_GETFL, 0);
+   if (flags >= 0)
+      fcntl(sck, F_SETFL, flags | O_NONBLOCK);
+
    return(sck);
 }
 #endif /* HAS_RFC2553 */
```

This is just a quick-and-dirty fix, as it sets nonblocking only on UDP sockets. Furthermore I haven't tested it on OS X and Windows.

Now the log looks like this: [dynamips_nonblock.txt](https://github.com/GNS3/gns3-gui/files/1167214/dynamips_nonblock.txt)

But maybe this issue can be fixed also by other means. Maybe we can skip the read, who we got an error on write() previously.

This patch fixes not only this issue, but also issue https://github.com/GNS3/gns3-gui/issues/2178

**Comment by ehlers** [comment]
Just found function m_fd_set_non_block in https://github.com/GNS3/dynamips/blob/master/common/utils.c#L397. That makes it easy to set nonblocking mode.

But I would change utils.c/line 402
`if ((flags = fcntl(fd,F_GETFL,0)) < 1)`
into
`if ((flags = fcntl(fd,F_GETFL,0)) < 0)`
as zero should be valid as well.

**Comment by grossmj** [comment]
Wow I am impressed, you have found the Ethernet switch bug we have been suspecting was there for 10 years ;) Now we still need to test on Windows and Mac. The biggest challenge is going to be able to compile on these platforms.

Thanks a lot! :+1:

**Comment by ehlers** [comment]
If you're happy with nonblocking I/O I can create a PR. I would activate nonblocking I/O only for UDP and Auto UDP sockets. That should minimize the number of changes. Or would you like to activate it for all connection types?

The drawback of nonblocking I/O is, that you can use it only after a select(). If there are some cases, where dynamips really wants to wait for some data (and block until it's get it), then changing to nonblocking will result in an EAGAIN error.

As I would use the already existing m_fd_set_non_block function to set nonblocking, compilation issues should be minimized. I can do it for Linux and Mac, but not for Windows. For Mac I would use gcc 4.9, the same Julien recommends in the dynamips readme.

**Comment by grossmj** [comment]
Yes please create a PR. Note that we only use UDP sockets in GNS3 (not the auto ones anymore).

We will take care of the Windows compilation.

Thanks!

**Comment by ehlers** [comment]
Created PR https://github.com/GNS3/dynamips/pull/81, tested on Linux and Mac OS X. Build instructions for OS X are included.

**Comment by grossmj** [comment]
This issue is fixed by the Dynamips version 0.2.17 (to be released soon).
